apps/heft/src/plugins/ProjectValidatorPlugin.ts

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as fs from 'fs';
+import { FileSystem, LegacyAdapters } from '@rushstack/node-core-library';
+
+import { HeftConfiguration } from '../configuration/HeftConfiguration';
+import { Constants } from '../utilities/Constants';
+import { ScopedLogger } from '../pluginFramework/logging/ScopedLogger';
+import { IHeftPlugin } from '../pluginFramework/IHeftPlugin';
+import { HeftSession } from '../pluginFramework/HeftSession';
+import { IHeftLifecycle } from '../pluginFramework/HeftLifecycle';
+
+const ALLOWED_HEFT_DATA_FOLDER_FILES: Set<string> = new Set<string>();
+const ALLOWED_HEFT_DATA_FOLDER_SUBFOLDERS: Set<string> = new Set<string>([Constants.buildCacheFolderName]);
+
+const PLUGIN_NAME: string = 'ProjectValidatorPlugin';
+
+export class ProjectValidatorPlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    heftSession.hooks.heftLifecycle.tap(PLUGIN_NAME, (heftLifecycle: IHeftLifecycle) => {
+      heftLifecycle.hooks.toolStart.tapPromise(PLUGIN_NAME, async () => {
+        const logger: ScopedLogger = heftSession.requestScopedLogger('project-validation');
+        await this._scanHeftDataFolderAsync(logger, heftConfiguration);
+      });
+    });
+  }
+
+  private async _scanHeftDataFolderAsync(
+    logger: ScopedLogger,
+    heftConfiguration: HeftConfiguration
+  ): Promise<void> {
+    // TODO: Replace this with a FileSystem API
+    let heftDataFolderContents: fs.Dirent[];
+    try {
+      // Use this instead of fs.promises to avoid a warning in older versions of Node
+      heftDataFolderContents = await LegacyAdapters.convertCallbackToPromise(
+        fs.readdir,
```

Comment:
Doesn't `FileSystem.readFolder()` do this?

Reply:
`FileSystem.readFolder()` doesn't differentiate between folders and files.